File: server/route/code_executor.py
import io
import sys
import traceback
from fastapi import APIRouter,Request
from json import loads
code_executor_route = APIRouter()

@code_executor_route.post("/execute")
async def home_execute(req:Request):
    code = await req.body()
    code = loads(code).get("code")
    output = io.StringIO()
    # 保存当前的 stdout
    old_stdout = sys.stdout
    sys.stdout = output
    result = ""
    try:
        exec(code)
        result = {"status": "success", "result": output.getvalue()}
    except Exception as e:
        sys.stdout = old_stdout
        result = {"status": "error", "result": traceback.format_exc()}
    finally:
        sys.stdout = old_stdout
    return result

File: app/src/main/routes/codeExecutor.ts
const { Router } = require('express');
const bodyParser = require('body-parser');
const { spawn } = require('child_process');

const router = Router();

// Parse incoming JSON
router.use(bodyParser.json());

// Define the '/execute' route to run Python code
router.post('/', (req, res) => {
    const code = req.body.code;  // The Python code to execute

    // Spawn a Python subprocess to execute the code
    const pyProcess = spawn('python', ['-c', code]);

    let output = '';
    let errorOutput = '';

    // Capture the Python process output (stdout)
    pyProcess.stdout.on('data', (data) => {
        output += data.toString();
    });

    // Capture any errors (stderr)
    pyProcess.stderr.on('data', (data) => {
        errorOutput += data.toString();
    });

    // When the Python process finishes
    pyProcess.on('close', (code) => {
        if (code === 0) {
            res.json({ status: 'success', result: output });
        } else {
            res.json({ status: 'error', result: errorOutput });
        }
    });
});

module.exports = router;
